collectors/mmrepquota.go

```diff
@@ -59,102 +68,226 @@ type QuotaMetric struct {
+			"GPFS group block quota", group_labels, nil),
+		GroupBlockLimit: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "limit_bytes"),
+			"GPFS group quota block limit", group_labels, nil),
+		GroupBlockInDoubt: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "in_doubt_bytes"),
+			"GPFS group quota block in doubt", group_labels, nil),
+		GroupFilesUsage: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "used_files"),
+			"GPFS group quota files used", group_labels, nil),
+		GroupFilesQuota: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "quota_files"),
+			"GPFS group files quota", group_labels, nil),
+		GroupFilesLimit: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "limit_files"),
+			"GPFS group quota files limit", group_labels, nil),
+		GroupFilesInDoubt: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "in_doubt_files"),
+			"GPFS group quota files in doubt", group_labels, nil),
+
 		logger: logger,
 	}
 }
 
 func (c *MmrepquotaCollector) Describe(ch chan<- *prometheus.Desc) {
-	ch <- c.BlockUsage
-	ch <- c.BlockQuota
-	ch <- c.BlockLimit
-	ch <- c.BlockInDoubt
-	ch <- c.FilesUsage
-	ch <- c.FilesQuota
-	ch <- c.FilesLimit
-	ch <- c.FilesInDoubt
+	ch <- c.FilesetBlockUsage
+	ch <- c.FilesetBlockQuota
+	ch <- c.FilesetBlockLimit
+	ch <- c.FilesetBlockInDoubt
+	ch <- c.FilesetFilesUsage
+	ch <- c.FilesetFilesQuota
+	ch <- c.FilesetFilesLimit
+	ch <- c.FilesetFilesInDoubt
+
+	ch <- c.UserBlockUsage
+	ch <- c.UserBlockQuota
+	ch <- c.UserBlockLimit
+	ch <- c.UserBlockInDoubt
+	ch <- c.UserFilesUsage
+	ch <- c.UserFilesQuota
+	ch <- c.UserFilesLimit
+	ch <- c.UserFilesInDoubt
+
+	ch <- c.GroupBlockUsage
+	ch <- c.GroupBlockQuota
+	ch <- c.GroupBlockLimit
+	ch <- c.GroupBlockInDoubt
+	ch <- c.GroupFilesUsage
+	ch <- c.GroupFilesQuota
+	ch <- c.GroupFilesLimit
+	ch <- c.GroupFilesInDoubt
 }
 
 func (c *MmrepquotaCollector) Collect(ch chan<- prometheus.Metric) {
 	level.Debug(c.logger).Log("msg", "Collecting mmrepquota metrics")
 	collectTime := time.Now()
 	timeout := 0
 	errorMetric := 0
-	metrics, err := c.collect()
-	if err == context.DeadlineExceeded {
-		timeout = 1
-		level.Error(c.logger).Log("msg", "Timeout executing mmrepquota")
-	} else if err != nil {
-		level.Error(c.logger).Log("msg", err)
-		errorMetric = 1
+	metrics := []QuotaMetric{}
+
+	typesToCollect := strings.Split(*configMmrepquotaTypes, ",")
+
+	results := make(chan MetricCollectionResult, len(typesToCollect)-1)
+
+	for _, quotaType := range typesToCollect {
+		quotaType = strings.TrimSpace(quotaType)
+		quotaArg := quotaTypeMap[quotaType]
+
+		// Collect quota types concurrently, place metrics on results channel as MetricCollectionResult
+		go func(quotaArg rune) {
+			metric, err := c.collect(fmt.Sprintf("-%c", quotaArg))
+			results <- MetricCollectionResult{Result: metric, Error: err}
+		}(quotaArg)
```

Comment:
How does this block on completion and not run to next code part of code before completion, is it the channel part?  I'm not a Go expert but when I've done go routines in a loop I usually have to add sync wait groups.

Reply:
Reading and writing to channels in Go is blocking by default. The for loop below will wait for all go routines to complete (or timeout) before code execution continues.
https://github.com/treydock/gpfs_exporter/blob/3590b822760c8a737279ddfebcb87f0d024a8940/collectors/mmrepquota.go#L222C2-L234C3